url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/io/DiskStorageManager.java

```diff
@@ -0,0 +1,423 @@
+
+public class DiskStorageManager implements IStorageManager {
+	private RandomAccessFile m_dataFile = null;
+	private RandomAccessFile m_indexFile = null;
+	private int m_pageSize = 0;
+	private int m_nextPage = -1;
+	/**
+	 * store the id of blank page
+	 */
+	private final TreeSet<Integer> m_emptyPages = new TreeSet<Integer>();
+	private final HashMap<Integer, Entry> m_pageIndex = new HashMap<Integer, Entry>();
+	/**
+	 * after init, it has the length of m_pageSize
+	 */
+	private byte[] m_buffer = null;
+
+	/**
+	 * ps should contain "Overwrite", "Filename", "Pagesize"
+	 * 
+	 * @param ps
+	 * @throws SecurityException
+	 * @throws NullPointerException
+	 * @throws IOException
+	 * @throws FileNotFoundException
+	 * @throws IllegalArgumentException
+	 */
+	public DiskStorageManager(PropertySet ps) throws SecurityException,
+			NullPointerException, IOException, FileNotFoundException,
+			IllegalArgumentException {
+		Object var;
+
+		// Open/Create flag.
+		boolean bOverwrite = false;
+		var = ps.getProperty("Overwrite");
+
+		if (var != null) {
+			if (!(var instanceof Boolean)) {
+				throw new IllegalArgumentException(
+						"Property Overwrite must be a Boolean");
+			}
+			bOverwrite = ((Boolean) var).booleanValue();
+		}
+
+		// storage filename.
+		var = ps.getProperty("FileName");
+
+		if (var != null) {
+			if (!(var instanceof String)) {
+				throw new IllegalArgumentException(
+						"Property FileName must be a String");
+			}
+
+			File indexFile = new File((String) var + ".idx");
+			File dataFile = new File((String) var + ".dat");
+
+			// check if files exist.
+			if (bOverwrite == false
+					&& (!indexFile.exists() || !dataFile.exists())) {
+				bOverwrite = true;
+			}
+
+			if (bOverwrite) {
+				if (indexFile.exists()) {
+					indexFile.delete();
+				}
+				if (dataFile.exists()) {
+					dataFile.delete();
+				}
+
+				boolean b = indexFile.createNewFile();
+				if (b == false) {
+					throw new IOException("Index file (" + (String) var
+							+ ".idx) cannot be opened.");
+				}
+
+				b = dataFile.createNewFile();
+				if (b == false) {
+					throw new IOException("Data file cannot be opened.");
+				}
+			}
+
+			m_indexFile = new RandomAccessFile(indexFile, "rw");
+			m_dataFile = new RandomAccessFile(dataFile, "rw");
+		} else {
+			throw new IllegalArgumentException(
+					"Property FileName was not specified.");
+		}
+
+		// find page size.
+		if (bOverwrite == true) {
+			var = ps.getProperty("PageSize");
+
+			if (var != null) {
+				if (!(var instanceof Integer)) {
+					throw new IllegalArgumentException(
+							"Property PageSize must be an Integer");
+				}
+				m_pageSize = ((Integer) var).intValue();
+				m_nextPage = 0;
+			} else {
+				throw new IllegalArgumentException(
+						"Property PageSize was not specified.");
+			}
+		} else {
+			try {
+				/*
+				 * This means page-size is stored at first
+				 */
+				m_pageSize = m_indexFile.readInt();
+			} catch (EOFException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 162, 202.